src/storage-area.cpp

```diff
@@ -0,0 +1,202 @@
+        if (file.good()) {
+            file.close();
+            return OrthancPluginErrorCode_Success;
+        }
+        OrthancPluginLogWarning(globals::context, "StorageCreateCallback: but write out appears bad");
+    }
+    return OrthancPluginErrorCode_FileStorageCannotWrite;
+}
+
+OrthancPluginErrorCode StorageReadWholeCallback(OrthancPluginMemoryBuffer64 *target,
+                                                const char *uuid,
+                                                OrthancPluginContentType type) {
+    auto path = GetPath(type,uuid);
+    // todo: is the buffer ready? without this call..
+    OrthancPluginCreateMemoryBuffer64(globals::context, target, fs::file_size(path));
+    std::fstream file(path, std::ios::in | std::ios::binary);
+    if(file.is_open()){
+        file.read((char*)target->data, target->size);
+        if(file.good()){
+            file.close();
+            return OrthancPluginErrorCode_Success;
+        }
+        OrthancPluginLogWarning(globals::context, "StorageReadWholeCallback: opened file, but couldn't read it");
+        return OrthancPluginErrorCode_StorageAreaPlugin;
+    }
+    return OrthancPluginErrorCode_InexistentFile;
+}
+
+OrthancPluginErrorCode StorageReadRangeCallback(OrthancPluginMemoryBuffer64 *target,
+                                                const char *uuid,
+                                                OrthancPluginContentType type,
+                                                uint64_t rangeStart) {
+    auto path = GetPath(type,uuid);
+    // todo: is the buffer ready? without this call..
+    OrthancPluginCreateMemoryBuffer64(globals::context, target, fs::file_size(path));
+    std::fstream file(path, std::ios::in | std::ios::binary);
+    if(file.is_open()){
+        file.seekg(rangeStart);
+        file.read((char*)target->data, target->size);
+        return OrthancPluginErrorCode_Success;
+    }
+    return OrthancPluginErrorCode_InexistentFile;
+}
+
+OrthancPluginErrorCode StorageRemoveCallback(const char *uuid, OrthancPluginContentType type) {
+    switch(type){
+        case OrthancPluginContentType_Dicom:
+            // todo: rewrite
+            const fs::path storage_root(globals::storage_location);
+            auto remove = [&](std::string group, std::string folder) {
+                fs::path path = fs::path(storage_root)
+                        .append(group)
+                        .append(folder)
+                        .append(uuid)
+                        .append(".DCM");
+                if(fs::exists(path)){
+                    fs::remove(path);
+                }
+            };
+            // todo: replace placeholders
+            std::string DOB_placeholder;
+            std::string PID_placeholder;
+            std::string SD_placeholder;
+            remove("/by-dob/", DOB_placeholder);
+            remove("/by-patient-id/", PID_placeholder);
+            remove("/by-study-date/", SD_placeholder);
```

Comment:
This is nearly the same code as above. There might be an elegant way of re-using code for the two sections. Just maybe not via a callback to call the callback